[OID4VCI] Extend realm UI configuration by OID4VCI attributes

forkimenjeckayang opened the pull request:
## Changes Made

### Frontend (Admin UI)
- **New OID4VCI Tab**: Added `RealmSettingsOid4vciTab.tsx` component with form fields for OID4VCI attributes
- **Tab Integration**: Integrated the new tab into `RealmSettingsTabs.tsx` with proper feature flag detection
- **Form Validation**: Implemented proper validation with minimum value requirements (min: 1) for both fields
- **Internationalization**: Added translation keys for all UI labels and help text
- **Routing**: Added route configuration for the new OID4VCI settings tab

### Backend Integration
- **Attribute Mapping**: Configured proper mapping between UI fields and realm attributes:
  - `vc.c-nonce-lifetime-seconds` - Controls nonce lifetime for OID4VCI flows
  - `preAuthorizedCodeLifespanS` - Controls pre-authorized code lifespan

### Testing
- **Comprehensive Test Suite**: Added `oid4vci-attributes.spec.ts` with full test coverage:
  - Feature flag detection (shows/hides tab based on OID4VCI feature availability)
  - Form field rendering and validation
  - Save functionality with correct attribute mapping
  - Required field validation
  - Accessibility compliance testing

### UI Features
- **Time Selector Components**: Utilizes existing TimeSelector components for intuitive time configuration
- **Form State Management**: Proper form state handling with save/revert functionality
- **Validation Feedback**: Real-time validation with appropriate error states
- **Feature Flag Support**: Gracefully handles environments where OID4VCI feature is disabled

## Configured Attributes

| UI Field | Realm Attribute | Description |
|----------|----------------|-------------|
| OID4VCI Nonce Lifetime | `vc.c-nonce-lifetime-seconds` | Lifetime of nonces used in OID4VCI flows |
| Pre-authorized Code Lifespan | `preAuthorizedCodeLifespanS` | Lifespan of pre-authorized codes in OID4VCI flows |

## Testing
- All tests pass with proper feature flag detection
- Form validation works correctly for both required fields and minimum values
- Accessibility compliance verified
- Integration with existing realm settings infrastructure confirmed

## Related
- Addresses ticket: `[OID4VCI] Extend realm UI configuration by OID4VCI attributes`
- Closes https://github.com/adorsys/keycloak-oid4vc/issues/63


<img width="1922" height="899" alt="image" src="https://github.com/user-attachments/assets/414d177b-9340-43e4-aa1f-42f094bc0750" />

IngridPuppet commented on js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties:
```diff
@@ -3515,3 +3515,9 @@ oid4vciEnabled=Enable OID4VCI
 oid4vciEnabledHelp=Enable this option to allow the client to request verifiable credentials from Keycloak's OID4VCI credential endpoint.
 noAccessPolicies=No access policies
 noAccessPoliciesInstructions=There haven't been configured any access policies yet. Click the button below to configure the first policy.
+oid4vciAttributes=OID4VCI Attributes
```
I notice other tabs do not use the title case. Maybe follow the trend?
```suggestion
oid4vciAttributes=OID4VCI attributes
```

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsTabs.tsx:
```diff
@@ -235,6 +236,34 @@ export const RealmSettingsTabs = () => {
+        });
+        // Remove the nested object
+        delete r.attributes["vc"];
+      }
```
Does this work in the other direction? Meaning if fields are pre-filled according to these values in the realm attributes.

forkimenjeckayang replied:
I checked the backend `RealmAdminResource` and related code. The realm attributes are always flat key-value pairs, never nested objects. The frontend form fields are mapped to these flat keys, so pre-filling works as expected. The flattening logic is only needed before saving, to ensure backend compatibility. No extra transformation is needed when loading data.
So this is expected 
```json
"attributes": {
  "vc.nonce-lifetime-seconds": "60",
  "preAuthorizedCodeLifespanS": "50"
}
```

Without flattening and saving , we will have somehthing like this
```json
"attributes": {
  "vc": {
    "nonce-lifetime-seconds": 60
  },
  "preAuthorizedCodeLifespanS": "50"
}
```

and update will fail

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsTabs.tsx:
```diff
@@ -463,6 +493,15 @@ export const RealmSettingsTabs = () => {
+              {...oid4vciTab}
+            >
+              <RealmSettingsOid4vciTab realm={realm!} save={save} />
+            </Tab>
```
I can't myself think of a better place for this configuration outside of Realm Settings. However, I'm afraid many won't agree to have a new tab because we won't be able to add tabs infinitely for any new protocol... Anyway, I'd side with you to keep this idea and if the point gets raised further, maybe think of a section under an existing tab.

forkimenjeckayang replied:
Noted @IngridPuppet

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -0,0 +1,106 @@
+  return (
+    <PageSection variant="light">
+      <FormPanel title={t("oid4vciAttributesSectionTitle")}>
+        <FormProvider {...form}>
```
The title is saying "Set OID4VC attributes". Maybe start creating sections like "General". I don't feel like other tabs are repeating their titles.

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -0,0 +1,106 @@
+                    id="oid4vciNonceLifetime"
+                    min={1}
+                    units={["second"]}
+                    value={field.value}
```
I notice the dropdown only contains "Seconds". Is it by choice? Other token lifetimes have more options. Check the "Tokens" tab.

forkimenjeckayang replied:
Yes, the backend for the OID4VCI attributes expects these values in seconds only (e.g., `preAuthorizedCodeLifespanS`, `c-nonce-lifetime-seconds`). There is no support for other time units, so the dropdown is intentionally limited to `seconds` to match backend requirements.

IngridPuppet replied:
Usually, the frontend shows seconds/minutes/hours just to be more user-friendly but does the conversion to seconds only under the hood. I truly doubt this is not the case with duration entries under the "Tokens" tab. Could you look into it? I also believe you won't have to do the conversion yourself given the `TimeSelector` construct available.

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -0,0 +1,106 @@
+                isDisabled={!formState.isDirty}
+              >
+                {t("save")}
+              </Button>
```
I thought at first that the "save" button wasn't working. Could you implement a user feedback post validation?

<img width="285" height="76" alt="image" src="https://github.com/user-attachments/assets/67450598-6951-42ef-853f-d5ca1e7e9554" />

IngridPuppet commented on js/apps/admin-ui/test/realm-settings/oid4vci-attributes.spec.ts:
```diff
@@ -0,0 +1,109 @@
+      await expect(oid4vciTab).toBeVisible();
+    } else {
+      await expect(oid4vciTab).toBeHidden();
+    }
```
"if/else" in tests is confusing because I don't know exactly what you're trying to test since both branches could be taken. What is the feature flag you wan't on or off? If on, should the tab be visible or hidden (exclusive or). 

By checking if the tab is visible, and then conditionally asserting if it is visible, you're not testing anything because the logic is tailored to pass whatsoever.

IngridPuppet commented on js/apps/admin-ui/test/realm-settings/oid4vci-attributes.spec.ts:
```diff
@@ -0,0 +1,109 @@
+      await expect(oid4vciTab).toBeHidden();
+    }
+  });
+});
```
Same as before. I'm not sure how to interpret the tests because of conditional branching on `isVisible`. Could you create the conditions for the tab to be visible and avoid the ifs?

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -18,9 +19,15 @@ export const RealmSettingsOid4vciTab = ({
+  // Show a global error notification if validation fails
+  const onError = () => {
+    addAlert(t("formValidationError"), AlertVariant.danger);
+  };
```
The "Email" tab does not fire a global alert, but a per-field feedback. Could you implement a similar behavior?

forkimenjeckayang replied:
Thank you for the feedback @IngridPuppet 
I reviewed the other Realm Settings tabs (such as Sessions and Tokens), which use time selectors and numeric fields similar to OID4VCI. These tabs do not show per-field inline error messages; instead, they rely on the Save button being disabled when the form is invalid.

The Email tab is a special case because it uses text inputs, which often require more granular validation feedback (e.g., for email formatting). For numeric/time fields, the established UX in Keycloak is to disable the Save button and not show inline errors.

For consistency and a unified user experience, I have aligned the OID4VCI tab with the Sessions tab and other similar tabs, rather than the Email tab.

Hope that's okay.?

IngridPuppet replied:
Definitely. Consistency was my only concern.

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -44,19 +51,27 @@ export const RealmSettingsOid4vciTab = ({
+                      .message
+                  }
+                </span>
+              )}
```
This does not seem to work...

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -0,0 +1,129 @@
+                name="attributes.vc.c-nonce-lifetime-seconds"
+                control={form.control}
+                rules={{ required: t("required"), min: 60 }}
+                render={({ field }) => (
```
There appears to be a problem with how this field is persisted. Do the following:
- Input a value and click save.
- Refresh the page.
- The field is blank (at least on my side).

Could you sort out the issue?

IngridPuppet commented on js/apps/admin-ui/test/realm-settings/oid4vci-attributes.spec.ts:
```diff
@@ -3,107 +3,111 @@ import { v4 as uuid } from "uuid";
+  await goToRealmSettings(page);
+  const oid4vciTab = page.getByTestId("rs-oid4vci-attributes-tab");
+  await page.waitForTimeout(1000);
+  isFeatureEnabled = (await oid4vciTab.count()) > 0;
```
A timeout is always almost unwanted because it is a sign of non-deterministic behavior. Could you wait for an element to appear on the page directly? It is generally preferable.

IngridPuppet commented on js/apps/admin-ui/test/realm-settings/oid4vci-attributes.spec.ts:
```diff
@@ -3,107 +3,111 @@ import { v4 as uuid } from "uuid";
+  } else {
+    // When feature is enabled, verify tab is visible
+    await expect(oid4vciTab).toBeVisible();
+  }
```
This update does not quite address the issue I previously raised about conditional branching in tests. Simply reading, I still don't know which branch to take and what is the behavior that you expect. 

I guess this whole testsuite does not matter if the OID4VCI feature is not on. Isn't that explicitly enabled in test configs? Edit: I checked and it is the case: 
https://github.com/adorsys/keycloak-oid4vc/blob/71f510d115a34ac78d56c1efd4cf5c4a9e9ef831/js/apps/keycloak-server/scripts/start-server.js#L61-L65

Then I suggest to simply assert that the tab is visible.

```suggestion
await expect(oid4vciTab).toBeVisible();
```
But you seem reluctant to. Could you communicate your rationale?

IngridPuppet commented on js/apps/admin-ui/test/realm-settings/oid4vci-attributes.spec.ts:
```diff
@@ -3,107 +3,111 @@ import { v4 as uuid } from "uuid";
+  if (!isFeatureEnabled) {
+    test.skip();
+    return;
+  }
```
Problematic, for the same reasons above.

IngridPuppet reviewed (changes requested):
I checked again and left further comments. Please could you check?

forkimenjeckayang commented:
Thanks for reviewing @IngridPuppet , i've addressed your suggestions and it's back for review again.
Just to highlight

I implemented manual hydration in the OID4VCI tab to ensure form values persist on refresh. The backend stores attributes with a flat structure `(vc.c-nonce-lifetime-seconds)`, but the form requires a nested structure `(vc: { 'c-nonce-lifetime-seconds': value })`. The generic form utilities couldn't handle this specific transformation, so i've added targeted hydration in the component to map the flat attributes to the nested form structure when the component mounts or receives new props.

All this is done bceause of how the attribuyte name has been defined. Other realm attributes are different.
<img width="567" height="308" alt="Screenshot from 2025-07-31 16-23-16" src="https://github.com/user-attachments/assets/4c8727de-3607-4d97-800b-868cab32c592" />

So, if we have to avoid all this, I would have to change the realm attribute definition. As you can see the pre-authorized code realm attribute has no issues.
Feedback on this too will be helpful on your next review

IngridPuppet commented on js/apps/admin-ui/src/realm-settings/RealmSettingsOid4vciTab.tsx:
```diff
@@ -19,14 +20,22 @@ export const RealmSettingsOid4vciTab = ({
-  // Show a global error notification if validation fails
-  const onError = () => {
-    addAlert(t("formValidationError"), AlertVariant.danger);
-  };
```
Oops! I was wrong... Some tabs fire a global alert like this. We need user feedback, that's the concern.

IngridPuppet commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java:
```diff
@@ -582,7 +582,8 @@ private VCIssuanceContext getVCToSign(List<OID4VCMapper> protocolMappers, Suppor
-    private void enforceKeyBindingIfProofProvided(VCIssuanceContext vcIssuanceContext) {
+    private void
+    enforceKeyBindingIfProofProvided(VCIssuanceContext vcIssuanceContext) {
         Proof proof = vcIssuanceContext.getCredentialRequest().getProof();
```
Certainly unwanted.

IngridPuppet reviewed (commented):
Just a few points...

IngridPuppet commented:
> Thanks for reviewing @IngridPuppet , i've addressed your suggestions and it's back for review again. Just to highlight
> 
> I implemented manual hydration in the OID4VCI tab to ensure form values persist on refresh. The backend stores attributes with a flat structure `(vc.c-nonce-lifetime-seconds)`, but the form requires a nested structure `(vc: { 'c-nonce-lifetime-seconds': value })`. The generic form utilities couldn't handle this specific transformation, so i've added targeted hydration in the component to map the flat attributes to the nested form structure when the component mounts or receives new props.
> 
> All this is done bceause of how the attribuyte name has been defined. Other realm attributes are different. <img alt="Screenshot from 2025-07-31 16-23-16" width="567" height="308" src="https://private-user-images.githubusercontent.com/104195313/473357733-4c8727de-3607-4d97-800b-868cab32c592.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQzOTg4MjYsIm5iZiI6MTc1NDM5ODUyNiwicGF0aCI6Ii8xMDQxOTUzMTMvNDczMzU3NzMzLTRjODcyN2RlLTM2MDctNGQ5Ny04MDBiLTg2OGNhYjMyYzU5Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgwNVQxMjU1MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MzYwODAzMzkyOGFmNDUxYmM0OTNiMDBjODg2ZjYzNGU1NWZjNDY1Yzg4M2M5NjVkYzBlODc0N2Y1N2U1ZDY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zsh7Q6IcC1M6nN-59nUymqbfnqg-t338ya4xkU9cALU">
> 
> So, if we have to avoid all this, I would have to change the realm attribute definition. As you can see the pre-authorized code realm attribute has no issues. Feedback on this too will be helpful on your next review

I cannot but agree that we're simply making our lives harder by adding these dots to attribute names. I noticed that the mechanism you implemented was also done for another pre-existing case. Given that we're only doing that for one attribute in this PR, I think it is okay to live with it. But if the task comes to handle more attributes, I'd definitely side with updating this naming style.

IngridPuppet commented on js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties:
```diff
@@ -2913,7 +2913,7 @@ policyProvider.client=Define conditions for your permissions where a set of one
 validatePasswordPolicy=Validate password policy
 registrationEmailAsUsername=Email as username
-webAuthnPolicyFormHelp=Policy for WebAuthn authentication. This one will be used by 'WebAuthn Register' required action and 'WebAuthn Authenticator' authenticator. Typical usage is, when WebAuthn will be used for the two-factor authentication.
+webAuthnPolicyFormHelp=Policy for WebAuthn authentication. This one will be used by 'Webauthn Register' required action and 'WebAuthn Authenticator' authenticator. Typical usage is, when WebAuthn will be used for the two-factor authentication.
```
This change is probably not intentional.

forkimenjeckayang replied:
definitely, will adjust before create a or for KC

forkimenjeckayang committed:
Extend realm UI configuration by OID4VCI attributes

Closes #39533

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
Merge branch 'main' into issue-39533

forkimenjeckayang committed:
update: adjust tests in oid4vci-attributes.spec.ts based on feature availability

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: directly check OID4VCI feature from server info in tests before running

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: address comment(s) by @IngridPuppet

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: skip tests when oid4vci feature is not enabled

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
fix: resolved merge conflicts

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: move oid4vc realm attributes setting to token tab

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: refactor tokens tab and restructure oid4vci attributes tests

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: apply review comments by @jonkoops

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
Merge branch 'main' into issue-39533

forkimenjeckayang committed:
update: refactored tests to use createTestBed()

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: apply changes by @jonkoops

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
Merge branch 'main' into issue-39533

forkimenjeckayang committed:
fix: resolved merge conflicts

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: address review comments

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: change test format

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: enable oid4vc feature in worfklows

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
Update .github/workflows/stability-js-ci.yml

Co-authored-by: Jon Koops <jonkoops@gmail.com>
Signed-off-by: forkimenjeckayang <104195313+forkimenjeckayang@users.noreply.github.com>

forkimenjeckayang committed:
fix: included required fields in tokens.ts as fix for CI tests

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update: address more review comments

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>
